### DISCUSSION TITLE
default values with optional attributes

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1399] [Q&A]
Hi, I have a question regarding the default values feature: what is the expected result when having an optional attribute with a default value?

For example, the Langium interface:

```ts
interface Node {
  // ...
  shape?: 'default' | 'cloud' = "default";
}
```

My thought was that it doesn't have to be parsed, but it would be required, not optional, in the TS interface:

```ts
interface Node {
  // ...
  shape: 'default' | 'cloud';
}
```

But it's not the case I guess.

**Comment by msujew** [comment]
The grammar type system mostly follows the same logic as TypeScript's type system. In TypeScript:

```ts
class X {
  readonly value?: string = 'default';
}
```

The `value` property still has type `string | undefined`, even though it is statically assigned. We do the same thing. You should be able to remove the `?` in the field declaration though, right?

**Reply by Yokozuna59** [reply]
> You should be able to remove the `?` in the field declaration though, right?

I don't think so. Removing the `?` would make the parser to expect the `shape` attribute, but it can't be parsed.

The above example is for mindmap diagram, which uses different syntax to identify the shape:

```
default shape
[square shape]
)cloud shape(
%% ...
```

**Reply by msujew** [reply]
I don't really follow. Why do you want a property on an AST elements that never gets parsed (so its effectively a system wide constant)?

**Reply by Yokozuna59** [reply]
How would we be able to know the type (shape) of the node without it? Our plan is to take the AST result, generate DB using the result, then render the diagram.

The approach I thought of is to override the parser, modify each node AST by checking the shape and setting the shape value.

**Reply by msujew** [reply]
Can you share the part of the grammar that's supposed to do the shape parsing? Maybe you could get away with a well constructed `ValueConverter` for this.

**Reply by Yokozuna59** [reply]
Here is it:

```
fragment MindmapNode returns MindmapNode:
  id=MINDMAP_NODE_ID?
  title=(
    MINDMAP_NODE_SQUARE_TITLE
    | MINDMAP_NODE_CIRCLE_TITLE
    | MINDMAP_NODE_ROUNDED_SQUARE_TITLE
    | MINDMAP_NODE_BANG_TITLE
    | MINDMAP_NODE_CLOUD_TITLE
    | MINDMAP_NODE_HEXAGON_TITLE
    | MINDMAP_NODE_DEFAULT_TITLE
  )
  EOL
  children+=MindmapNode*
;

// it's an override rule: /([^\n\r()[{]+)/;
terminal MINDMAP_NODE_ID: /mindmap-node-id/;
terminal MINDMAP_NODE_SQUARE_TITLE: /\[("[^"]+"|[^\]]+)]/;
terminal MINDMAP_NODE_CIRCLE_TITLE:  /\(\(("[^"]+"|[^)]+)\)\)/;
terminal MINDMAP_NODE_ROUNDED_SQUARE_TITLE: /\(("[^"]+"|[^)]+)\)/;
terminal MINDMAP_NODE_BANG_TITLE: /\)\)("[^"]+"|[^(]+)\(\(/;
terminal MINDMAP_NODE_CLOUD_TITLE: /\)("[^"]+"|[^(]+)\(/;
terminal MINDMAP_NODE_HEXAGON_TITLE: /{{("[^"]+"|[^}]+)}}/;
// it's an override rule: /([^\n\r]+)/;
terminal MINDMAP_NODE_DEFAULT_TITLE: /mindmap-node-default-title/; 
```

Each node can have an id, a title, or both. If it's a default node, it doesn't have a title:

```
cool      %% id:cool,            shape:default
cool[wow] %% id:cool, title:wow, shape:square
[wow]     %%          title:wow, shape:square
```

**Reply by msujew** [reply]
Oh, I see, so a computed property effectively? That's definitely not supported right now, and I'm not sure whether it's in scope of Langium at all.

**Reply by Yokozuna59** [reply]
Yes. Well, that's the syntax of the [mindmap](http://mermaid.js.org/syntax/mindmap.html) diagram, which is similar to the [flowchart](http://mermaid.js.org/syntax/flowchart.html) and [block](http://mermaid.js.org/syntax/block.html) diagrams.
